operator/controllers/resources/user_controller.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gravitational/teleport/api/client"
+	resourcesv2 "github.com/gravitational/teleport/operator/apis/resources/v2"
+	"github.com/gravitational/trace"
+)
+
+// UserReconciler reconciles a User object
+type UserReconciler struct {
+	kclient.Client
+	Scheme         *runtime.Scheme
+	TeleportClient *client.Client
+}
+
+//+kubebuilder:rbac:groups=resources.teleport.dev,resources=users,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=resources.teleport.dev,resources=users/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=resources.teleport.dev,resources=users/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the User object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return ResourceBaseReconciler{
+		Client:         r.Client,
+		DeleteExternal: r.Delete,
+		UpsertExternal: r.Upsert,
+	}.Do(ctx, req, &resourcesv2.User{})
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *UserReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&resourcesv2.User{}).
+		Complete(r)
+}
+
+func (r *UserReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	return r.TeleportClient.DeleteUser(ctx, obj.GetName())
+}
```

Comment:
The delete does not check the unicity of the link between the CR and the teleport resource. The following scenario demonstrates a drift between the real state and the desired state.

```
$ cat user.yaml
apiVersion: resources.teleport.dev/v2
kind: User
metadata:
  name: eva
spec:
  roles:
    - yolo
  traits:
    logins:
      - root

$ kubectl create -f user.yaml -n ns1
# the user is now present in Kubernetes and Teleport

$ kubectl create -f user.yaml -n ns2
# the user is now present twice in Kubernetes and once in Teleport

$ kubectl delete -f user.yaml -n ns1
# the user is now present once in Kubernetes and is not in teleport
```

Such drift will stay until a new even happen on the user, the controller is restarted, or until a full resync (every 10 hours).

This raises 2 questions:
- Proof of ownership from the operator (if the user was here before the operator, we create and delete it, shoult it be deleted from teleport as well, @Joerger  is also raising the question here: https://github.com/gravitational/teleport/pull/13331/files#r910388983)
- Proof of 1-1 mapping between the kube ressource and the teleport resource. What happens if there are multiple operators fighting for the same resource? Multiple resources trying to represent the same user?

I think it might be saner for the operator to:
- label the namespace the resource is coming from (name + namespace should be a unique idenfier in a Kube cluster)
- label the resources with the release name, or at least a unique key we give to the operator (ensure we can ensure unicity across multiple Kubernetes clusters, `external-dns` does this and it saved my life a couple of times)
- refuse to interact with a teleport resources incorrectly labeled

It means users wanting to "adopt" existing Teleport resources in Kubernetes will have to label them accordingly. This is an extra step but might be safer.

This would also allow us to be idempotent and clean resources on startup even if a finalizer failed (if the Kubernetes cluster was down/destroyed for some reason) and we have no trace of the Kube CR.

Reply:
The first part should be solved, because we added a restriction to the namespaces the controller is watching.
So, in this case `kubectl create -f user.yaml -n ns2` would either not work or create the resource in another Teleport Cluster

The other topic seems a lot more tricky
What should we do when states diverge? either because it already existed
a) merge the properties
b) only update or create **if the resource has our origin label** and return an error otherwise
c) always update the resource, no matter what

Right now we have c)
I kinda of like option b) (proof of ownership or ownership handoff)

@Joerger I think your comment above is about this issue, so let's discuss it in this thread
https://github.com/gravitational/teleport/pull/13331#discussion_r910388983
If not, please let me know

@r0mant any strong opinion on this?

Reply:
I don't have a strong preference either way, but if we go with the current approach (c), then an update should always result in the origin label getting added, and should not update ignored fields I think? In my comment it shows how `created_by` gets set to a zero value in this case, but in the normal operator create -> operator update case, the ignored fields like `created_by` get properly ignored.

Reply:
I gave a try in the PR https://github.com/gravitational/teleport/pull/14099

- Following a discussion with @r0mant: I implemented solution b
- The PR adds a user feedback through the use of status.conditions
- It also adds custom resource definition status generation (with a couple of ugly fixes to locate packages, I'll try to revamp this later) from go types in `crdgen`

> Also, note that currently (as far as I can tell) there is no way to create a new user through the operator which can be logged in with. You are forced to `kubectl exec -- tctl users add user --roles=roles` to get the invite token, and then if you want you can update the user afterwards with the operator. Currently this flow leads to the problematic case described above.

Users can (and should?) use UI/CLI to get reset links on resources created by the operator, as for the Terraform provider. Either by doing `tctl users reset myuser` or clicking on the `Reset password` button on the Users UI tab.